packages/frontend/component/src/components/attachment-viewer/viewer.tsx

```diff
@@ -0,0 +1,417 @@
+        `[data-index="${id}"] canvas`
+      );
+      if (!canvas) return;
+      if (canvas.dataset.rendered) return;
+
+      // TODO(@fundon): improve
+      const ctx = canvas.getContext('2d');
+      if (ctx) {
+        ctx.putImageData(imageData, 0, 0);
+        canvas.dataset.rendered = 'true';
+      }
+    },
+    [scrollerRef]
+  );
+
+  const renderThumbnail = useCallback(
+    (id: number, imageData: ImageData) => {
+      const el = thumbnailsScrollerRef.current;
+      if (!el) return;
+
+      const canvas: HTMLCanvasElement | null = el.querySelector(
+        `[data-index="${id}"] canvas`
+      );
+      if (!canvas) return;
+      if (canvas.dataset.rendered) return;
+
+      // TODO(@fundon): improve
+      const ctx = canvas.getContext('2d');
+      if (ctx) {
+        ctx.putImageData(imageData, 0, 0);
+        canvas.dataset.rendered = 'true';
+      }
+    },
+    [thumbnailsScrollerRef]
+  );
+
+  const onScroll = useCallback(() => {
+    const el = scrollerRef.current;
+    if (!el) return;
+
+    const { scrollTop, scrollHeight } = el;
+
+    setDocInfo(info => {
+      const cursor = Math.ceil(scrollTop / (scrollHeight / info.total));
+      // thumbnailsScrollerHandleRef.current?.scrollToIndex(cursor)
+      return {
+        ...info,
+        cursor,
+      };
+    });
+    // }, [scrollerRef, thumbnailsScrollerHandleRef]);
+  }, [scrollerRef]);
+
+  const onSelect = useCallback(
+    (index: number) => {
+      scrollerHandleRef.current?.scrollToIndex(index);
+      setDocInfo(info => ({ ...info, cursor: index }));
+    },
+    [scrollerHandleRef]
+  );
+
+  const updateMainVisibleRange = useMemo(
+    () => debounce(setMainVisibleRange, 233, { leading: true, trailing: true }),
+    [setMainVisibleRange]
+  );
+
+  const updateThumbnailsVisibleRange = useMemo(
+    () =>
+      debounce(setThumbnailsVisibleRange, 233, {
+        leading: true,
+        trailing: true,
+      }),
+    [setThumbnailsVisibleRange]
+  );
+
+  // useEffect(() => {
+  //   const el = viewerRef.current;
+  //   if (!el) return;
+
+  //   return observeResize(el, entry => {
+  //     console.log(entry);
+  //   });
+  // }, []);
+
+  useEffect(() => {
+    const { startIndex, endIndex } = mainVisibleRange;
+    let index = startIndex;
+    for (; index < endIndex + 1; index++) {
+      post(MessageOp.Render, { index, kind: 'page' });
+    }
+  }, [mainVisibleRange, post]);
+
+  useEffect(() => {
+    const { startIndex, endIndex } = thumbnailsVisibleRange;
+    let index = startIndex;
+    for (; index < endIndex + 1; index++) {
+      post(MessageOp.Render, { index, kind: 'thumbnail' });
+    }
+  }, [thumbnailsVisibleRange, post]);
+
+  useEffect(() => {
+    workerRef.current = new Worker(
+      /* webpackChunkName: "pdf.worker" */ new URL(
+        './worker/worker.ts',
+        import.meta.url
+      )
+    );
+
+    async function process({ data }: MessageEvent<MessageData>) {
```

Comment:
how about using https://github.com/Jack-Works/async-call-rpc to abstract the message passing logic?

Reply:
Good idea. 

We can add it into https://github.com/toeverything/AFFiNE/pull/8615.